### PR TITLE
Create engagement tfact models

### PR DESCRIPTION
**Pull request by rachellougee**
### What are the relevant tickets?

https://github.com/mitodl/ol-data-platform/issues/1515

### Description (What does it do?)

Creating `tfact_course_navigation_events`, `tfact_problem_events` and `tfact_video_events`

### How can this be tested?

dbt build --select tfact_course_navigation_events
dbt build --select tfact_video_events
dbt build --select tfact_problem_events

### CODE REVIEW

File: src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        , coalesce(current_tab, starting_tab, starting_url) as starting_position
+        , coalesce(ending_tab, ending_url) as ending_position
+        , current_tab
+        , tab_count
+        , event_timestamp
+    from edxorg_navigation_events
+)
+
+select
+     {{ generate_hash_id('cast(user_id as varchar) || platform') }} as user_id
```
Comment (rachellougee) [inline src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql:173]: FYI, this might change depends on the dim_users that Kate is working on

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 2]
```diff
@@ -76,3 +76,150 @@ models:
...
+      the navigation event occurred
+    tests:
+    - not_null
+  - name: courserun_readable_id
+    description: string, foreign key to dim_course_content, referencing the edX course
+      run ID
+    tests:
+    - not_null
+  - name: event_type
+    description: string, type of the navigation event, e.g. edx.ui.lms.jump_nav.selected,
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:99]: Can we link to the open edX docs for event_type? I presume there is a complete & up-to-date list of events there.

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 3]
```diff
@@ -76,3 +76,150 @@ models:
...
+  - name: video_duration
+    description: float, the length of the video, in seconds.
+  - name: video_position
+    description: number, the seconds into the video where the event occurred. May
+      be Null for `load_video` or `seek_video` events.
+  - name: starting_position
+    description: number, the seconds into the video where the user was before seeking
+      the video. For very small seek_video events, the `old_time` may not be recorded
+      in the residential dataset.
+  - name: ending_position
```
Comment (rachellougee) [inline src/ol_dbt/models/dimensional/_dim__models.yml:164]: The naming for these columns might be debatable. But I am open to changing them if you think they're confusing.

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 4]
```diff
@@ -76,3 +76,150 @@ models:
...
+      run ID
+    tests:
+    - not_null
+  - name: event_type
+    description: string, type of the navigation event, e.g. edx.ui.lms.jump_nav.selected,
+      edx.ui.lms.link_clicked, edx.ui.lms.outline.selected, edx.ui.lms.sequence.next_selected,
+      edx.ui.lms.sequence.previous_selected, edx.ui.lms.sequence.tab_selected, edx.ui.lms.video.played
+    tests:
+    - not_null
+  - name: block_id
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:104]: Can we have a couple of examples? If possible, one with a uuid and one with a readable id (courses authored in XML sometimes have readable block_ids.

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 5]
```diff
@@ -76,3 +76,150 @@ models:
...
+      the course
+    tests:
+    - not_null
+  - name: platform_id
+    description: string, foreign key to dim_platform, referencing the platform where
+      the navigation event occurred
+    tests:
+    - not_null
+  - name: courserun_readable_id
+    description: string, foreign key to dim_course_content, referencing the edX course
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:94]: Just checking: Do courses and blocks both appear in dim_course_content?

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 6]
```diff
@@ -76,3 +76,150 @@ models:
...
+      edx.ui.lms.sequence.previous_selected, edx.ui.lms.sequence.tab_selected, edx.ui.lms.video.played
+    tests:
+    - not_null
+  - name: block_id
+    description: string, foreign key to dim_course_content, referencing the block
+      ID of the subsection that is being navigated by the user. Populated only for
+      events that are related to course subsections.
+  - name: starting_position
+    description: string, the starting position where the user is navigating from,
+      either a page URL or the tab in the unit navigation bar.
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:110]: Can you include an example of "the tab in the unit navigation bar" -- assuming it's not a URL

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 7]
```diff
@@ -76,3 +76,150 @@ models:
...
+  - name: ending_position
+    description: string, the ending position where the user is navigating to, either
+      a page URL or the tab in the unit navigation bar. If ending_position is NULL
+      but starting_position is not NULL, it means the user navigates from the first
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:114]: This is a really weird edge case. Is there some documentation you can reference here?

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 8]
```diff
@@ -76,3 +76,150 @@ models:
...
+      the video event occurred
+    tests:
+    - not_null
+  - name: courserun_readable_id
+    description: string, foreign key to dim_course_content, referencing the edX course
+      run ID
+    tests:
+    - not_null
+  - name: event_type
+    description: string, type of the video event. Possible values are load_video,
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:146]: add link to docs

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 9]
```diff
@@ -76,3 +76,150 @@ models:
...
+      pause_video, play_video, seek_video, speed_change_video, stop_video, and video_hide_transcript.
+    tests:
+    - not_null
+  - name: video_id
+    description: string, foreign key referencing a substring of the video block ID
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:151]: can we have an example? 

I'd also like to know where this id is stored in the edx application, but I'm not sure it's needed in this description string.

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 10]
```diff
@@ -76,3 +76,150 @@ models:
...
+      in the dim_course_content table.
+    tests:
+    - not_null
+  - name: video_duration
+    description: float, the length of the video, in seconds.
+  - name: video_position
+    description: number, the seconds into the video where the event occurred. May
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:158]: why not a float, like the length?

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 11]
```diff
@@ -76,3 +76,150 @@ models:
...
+      be Null for `load_video` or `seek_video` events.
+  - name: starting_position
+    description: number, the seconds into the video where the user was before seeking
+      the video. For very small seek_video events, the `old_time` may not be recorded
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:162]: `old_time` ?

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 12]
```diff
@@ -76,3 +76,150 @@ models:
...
+      the problem event occurred
+    tests:
+    - not_null
+  - name: courserun_readable_id
+    description: string, foreign key to dim_course_content, referencing the edX course
+      run ID
+    tests:
+    - not_null
+  - name: event_type
+    description: string, type of the problem event. Possible values are problem_check,
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:198]: add link to docs

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 13]
```diff
@@ -76,3 +76,150 @@ models:
...
+      , showanswer
+    tests:
+    - not_null
+  - name: problem_id
+    description: string, foreign key referencing the problem block ID in the dim_course_content
+      table.
+    tests:
+    - not_null
+  - name: attempt
+    description: number, number indicating which attempt this was
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:208]: Do attempts start with 1 or zero?

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 14]
```diff
@@ -76,3 +76,150 @@ models:
...
+  - name: answers
+    description: json, student answers to this problem in problem_id and internal
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:210]: learner or student?

File: src/ol_dbt/models/dimensional/_dim__models.yml [hunk 15]
```diff
@@ -76,3 +76,150 @@ models:
...
+      answer pair. For multiple questions, it lists every question and answer.
```
Comment (pdpinch) [inline src/ol_dbt/models/dimensional/_dim__models.yml:211]: Can we have some examples? I don't understand this structure.

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 5 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml:94]: Yes, the courses are the top level of course structure. Here is the example you can see https://courses.mitxonline.mit.edu/api/course-structure/v0/course-v1:MITxT+14.740x+1T2025/

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 2 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml:99]: Yes. I added the link. The link is very long and might not be readable in the description.

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 6 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml:112]: I updated the description. Hopefully its clear now

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 7 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml:116]: This is actually not a edge case. It is common when user nagivates across subsections. Here is the open edx doc 
https://edx.readthedocs.io/projects/devdata/en/latest/internal_data_formats/tracking_logs/student_event_types.html#example-edx-ui-lms-sequence-next-selected-events

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 9 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml:155]: This video_id is just an hash code in the block ID `block-v1:{org}+{course}+{run}type@video@{hash code}.`. This ID is different from whats stored in the edx application `edx_video_id`
```
  "block-v1:MITxT+14.740x+1T2025+type@video+block@52daa7bf99844138a4f06813e95b1917": {
    "category": "video",
    "children": [],
    "metadata": {
      "display_name": "The MicroMasters in DEDP",
      "download_track": true,
      "download_video": false,
      "edx_video_id": "96d1b643-cab3-4b76-8c8d-f54fc69b23d3",
```

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 4 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml:106]: This is a block ID formatted like block-v1:{org}+{course}+{run}type@vertical+block@{hash code}. I am not sure about readable block_ids. Do we really need to provide examples for these cases in the description?

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 10 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml]: It is float. `number` is used in the open edx doc, so it's a copy-paste issue.

File: src/ol_dbt/models/dimensional/_dim__models.yml (hunk 13 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/_dim__models.yml]: attempt starts with 1

**Comment by rachellougee** [comment]
>  I think it looks like these tables don't have any unique contraints/checks which we may need to add. And it looks like they don't have primary key's which I think we could potentially add as well.

I added the uniqueness check. Given the nature of the tracking log, uniqueness is by the combination of almost every columns. I'm not sure how useful it will be, but I included it anyway.

For the primary key, are you suggesting that we should generate a surrogate key in dbt?

**Comment by rachellougee** [comment]
> I think that's pretty typical for fact tables as they are made up of a lot of foreign keys. Yes, a surrogate key would work

It is typical to have a surrogate key for fact table. But given the nature of tracking log, we would need to generate surrogate key that is a combination of all the columns "user_id", "courserun_readable_id", "event_type", "block_id", "event_json", "event_timestamp". This is quite expensive, and adding unique check, it has already increased dbt build time. So before I go about adding another surrogant key, it’s important to understand the use cases and benefits of this key

@blarghmatey do you have any thought?

**Comment by blarghmatey** [comment]
For all of the foreign key fields we should identify those with a `_fk` suffix. Also, looking at the logic in the models, it doesn't seem like it's using the primary key field from the dimensional models? We should be deliberate about the creation of surrogate keys as a `_pk` field in the dimensional models, and then referencing those as the `_fk`. This is especially important for dimensions that are using a Type 2 SCD (Slowly Changing Dimension).

For surrogate key generation we might want to consider using the surrogate_key macro from the dbt_utils package https://github.com/dbt-labs/dbt-utils/tree/1.3.0/?tab=readme-ov-file#generate_surrogate_key-source (a useful reference: https://docs.getdbt.com/blog/kimball-dimensional-model#step-4-create-the-surrogate-key)

**Comment by blarghmatey** [comment]
As discussed in our meeting, for cases where we don't have an existing dimension to reference as a foreign key (e.g. course_key) we should try to avoid using `_id` as a suffix to avoid confusion, but not have that be a blocker on getting the facts implemented.

**Comment by rachellougee** [comment]
@blarghmatey @KatelynGit @quazi-h if any of you guys can take a look and approve the PR, that would be great

File: src/ol_dbt/models/dimensional/tfact_video_events.sql [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+        , event_timestamp
+    from edxorg_video_events
+)
+
+select distinct
+    platform
+    , openedx_user_id
+    , courserun_readable_id
+    , event_type
+    , video_id
```
Comment (blarghmatey) [inline src/ol_dbt/models/dimensional/tfact_video_events.sql:174]: Is this the block_id for the video? Should it be an FK to the dim_content table?

File: src/ol_dbt/models/dimensional/tfact_problem_events.sql [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    from edxorg_problem_events
+)
+
+select
+
+    platform
+    , openedx_user_id
+    , courserun_readable_id
+    , event_type
+    , problem_block_id
```
Comment (blarghmatey) [inline src/ol_dbt/models/dimensional/tfact_problem_events.sql:170]: Should this be an FK to the `dim_content` table?

File: src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+        , event_timestamp
+    from edxorg_navigation_events
+)
+
+select
+    platform
+    , openedx_user_id
+    , courserun_readable_id
+    , event_type
+    , block_id
```
Comment (blarghmatey) [inline src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql:174]: Should this be an FK to the `dim_content`?

File: src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+        , tab_count
+        , event_timestamp
+    from edxorg_navigation_events
+)
+
+select
+    platform
+    , openedx_user_id
+    , courserun_readable_id
+    , event_type
```
Comment (blarghmatey) [inline src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql:173]: Not for this PR, but something to keep in mind going forward, is that we should have something like a `dim_event` to give us an Enum of event types and what they mean.

Comment (blarghmatey) [inline src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql:173]: Not for this PR, but something to keep in mind going forward, is that we should have something like a `dim_event` to give us an Enum of event types and what they mean.

File: src/ol_dbt/models/dimensional/tfact_video_events.sql (hunk 16 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/tfact_video_events.sql:174]: I can do all these in this PR but I was going to create a separate PR for dim_course_content, dim_video and dim_problem

File: src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql (hunk 19 above)
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/tfact_course_navigation_events.sql:173]: I will add dim_event as part of my other PR for dim_ tables

**Review by blarghmatey — APPROVED** [review]
